core/src/main/java/org/apache/iceberg/avro/SchemaToType.java

```diff
@@ -105,13 +106,27 @@ public Type record(Schema record, List<String> names, List<Type> fieldTypes) {
 
   @Override
   public Type union(Schema union, List<Type> options) {
-    Preconditions.checkArgument(AvroSchemaUtil.isOptionSchema(union),
-        "Unsupported type: non-option union: %s", union);
-    // records, arrays, and maps will check nullability later
-    if (options.get(0) == null) {
-      return options.get(1);
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      // Optional simple union
+      // records, arrays, and maps will check nullability later
+      if (options.get(0) == null) {
+        return options.get(1);
+      } else {
+        return options.get(0);
+      }
     } else {
-      return options.get(0);
+      // Complex union
+      List<Types.NestedField> newFields = new ArrayList<>();
+      newFields.add(Types.NestedField.required(allocateId(), "tag", Types.IntegerType.get()));
+
+      int tagIndex = 0;
+      for (Type type : options) {
+        if (type != null) {
+          newFields.add(Types.NestedField.optional(allocateId(), "field" + tagIndex++, type));
+        }
```

Comment:
Can you add the reasoning behind why the field names `tag` and `field${index}` were chosen when converting to struct? Spark's Avro datasources uses `member${index}` while Hive's `extract_union` UDF uses `tag_${index}`.

Reply:
The schema of struct converted from complex union is chosen to be consistent with Trino representation as implemented in this PR: https://github.com/trinodb/trino/pull/3483